File: apps/server/src/modules/loyalty/loyalty.controller.ts
import type { Request, Response } from 'express';
import { loyaltyService } from './loyalty.service.js';
import { asyncHandler } from '../../utils/asyncHandler.js';
import { ok } from '../../utils/ApiResponse.js';
import { ApiError } from '../../utils/ApiError.js';

export const loyaltyController = {
  me: asyncHandler(async (req: Request, res: Response) => {
    const data = await loyaltyService.getProfile(req.user!.id);
    return ok(res, data, 'Loyalty profile');
  }),

  applyReferral: asyncHandler(async (req: Request, res: Response) => {
    const code = req.body?.code;
    if (!code) throw ApiError.badRequest('code is required');
    const data = await loyaltyService.applyReferral(req.user!.id, code);
    return ok(res, data, 'Referral applied');
  }),
};

File: apps/guzo-api/src/main/java/et/guzo/web/controller/LoyaltyController.java
package et.guzo.web.controller;

import et.guzo.common.ApiException;
import et.guzo.common.ApiResponse;
import et.guzo.security.RoleChecker;
import et.guzo.security.SecurityUtil;
import et.guzo.service.LoyaltyService;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/loyalty")
@RequiredArgsConstructor
public class LoyaltyController {

    private final LoyaltyService loyaltyService;

    @GetMapping("/me")
    public ApiResponse<Map<String, Object>> me() {
        if (!RoleChecker.hasAnyRole(SecurityUtil.requireUser(), "CUSTOMER")) {
            throw ApiException.forbidden("Customer access required");
        }
        return ApiResponse.ok(loyaltyService.getProfile(SecurityUtil.requireUser().getId()));
    }

    @PostMapping("/referral")
    public ApiResponse<Map<String, Object>> applyReferral(@RequestBody Map<String, String> body) {
        if (!RoleChecker.hasAnyRole(SecurityUtil.requireUser(), "CUSTOMER")) {
            throw ApiException.forbidden("Customer access required");
        }
        return ApiResponse.ok(loyaltyService.applyReferral(SecurityUtil.requireUser().getId(), body.get("code")));
    }
}
